stdlib/public/core/String.swift

```diff
@@ -925,91 +925,57 @@ extension String {
 }
 
 // Index conversions
-extension String.Index {
-  /// Construct the position in `characters` that corresponds exactly to
-  /// `unicodeScalarIndex`. If no such position exists, the result is `nil`.
-  ///
-  /// - Precondition: `unicodeScalarIndex` is an element of
-  ///   `characters.unicodeScalars.indices`.
-  public init?(
-    _ unicodeScalarIndex: String.UnicodeScalarIndex,
-    within characters: String
-  ) {
-    if !unicodeScalarIndex._isOnGraphemeClusterBoundary {
-      return nil
-    }
-    self.init(_base: unicodeScalarIndex)
-  }
-
-  /// Construct the position in `characters` that corresponds exactly to
-  /// `utf16Index`. If no such position exists, the result is `nil`.
-  ///
-  /// - Precondition: `utf16Index` is an element of
-  ///   `characters.utf16.indices`.
-  public init?(
-    _ utf16Index: String.UTF16Index,
-    within characters: String
-  ) {
-    if let me = utf16Index.samePosition(
-      in: characters.unicodeScalars
-    )?.samePosition(in: characters) {
-      self = me
-    }
-    else {
-      return nil
-    }
-  }
-
-  /// Construct the position in `characters` that corresponds exactly to
-  /// `utf8Index`. If no such position exists, the result is `nil`.
-  ///
-  /// - Precondition: `utf8Index` is an element of
-  ///   `characters.utf8.indices`.
-  public init?(
-    _ utf8Index: String.UTF8Index,
-    within characters: String
-  ) {
-    if let me = utf8Index.samePosition(
-      in: characters.unicodeScalars
-    )?.samePosition(in: characters) {
-      self = me
-    }
-    else {
-      return nil
-    }
-  }
-
-  /// Returns the position in `utf8` that corresponds exactly
-  /// to `self`.
-  ///
-  /// - Precondition: `self` is an element of `String(utf8).indices`.
+extension String {
+  // TODO: swift-3-indexing-model - add docs
   @warn_unused_result
-  public func samePosition(
-    in utf8: String.UTF8View
-  ) -> String.UTF8View.Index {
-    return String.UTF8View.Index(self, within: utf8)
+  public func convert(
+    index index: String.UTF8Index,
+    from view: String.UTF8View
+  ) -> String.Index? {
```

Comment:
We need to think about this API.  Do we really need both buffers available to perform the conversion?

Something to think about: we know that the input index has to be valid with regards to the view it corresponds to.  We know that the destination view has to have the same prefix up to the index as the source view (that is, the Unicode scalars should match exactly).

I feel like given these constraints, we can keep the old API signature.

My hesitation with changing this API is primarily caused by the existing complexity of it.  I would like to avoid making it more confusing or cumbersome to use if at all possible.